README.md
# React Kaltura Resumable JS

Uploader with [ReactResumableJS](https://www.npmjs.com/package/react-resumable-js) to upload file into Kaltura

[![NPM](https://nodei.co/npm/react-kaltura-resumablejs.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/react-kaltura-resumablejs/)

[![CircleCI](https://circleci.com/gh/Artear/ReactKalturaResumableJS.svg)](https://circleci.com/gh/Artear/ReactKalturaResumableJS) [![CircleCI](https://circleci.com/gh/Artear/ReactKalturaResumableJS.svg?style=shield)](https://circleci.com/gh/Artear/ReactKalturaResumableJS)

### Install
`npm i react-kaltura-resumablejs`

### Node Version
- LTS

### Options

- server: http path to upload the file. Ex: http://kaltura.com/api_v3
- ks: Kaltura secret key from Kaltura.
- uploaderID: Container ID to render the ReactResumableJS. Ex: "video-uploader",
- filetypes: Set the  video extensions to upload. Ex: ['mp4','avi'].
- chunksize: Set in KB the size to upload the file in chunks. Ex: 1024.
- simultaneousUploads: amount of upload at same time. Ex: 1.
- onError: Catch the upload error. Must be a function. Ex: (error) => { return error; }.
- onSuccess: Catch the file success when finish the upload. Ex: (success) => { return success}
- textLabel: "Uploaded files",
- fileAddedMessage: "Started",
- completedMessage: "Complete!:"
- disableInput Boolean to disable or enable input file. Send true to disable, false otherwise.
- maxFileSize The allowed file size for upload. Is expressed in bytes. Default is: 104857600 (100mb).
- maxFiles Indicates how many files can be uploaded in a single session. Valid values are any positive integer and undefined for no limit. (Default: undefined)
- onMaxFileSizeErrorCallback Usefull to use with the above param, and take the exception for use as you want.
- maxFilesErrorCallback A function which displays the please upload n file(s) at a time message. (Default: displays an alert box with the message Please n one file(s) at a time.)
- fileAccept Set the file accept.
### Changelogs
[Changelogs](changelog.md)

### Example

```javascript
export default class Example extends React.Component {

    constructor(props) {
        super(props);
    }
    
    render() {

        return (

            <div>
               <ReactKalturaResumableJs
                   server="http://www.kaltura.com/api_v3"
                   ks="0_asdfafdvsdbsdgsdgsdgs"
                   uploaderID="myKalturaUploader"
                   onFileSuccess={(success) => { console.log(success); }}
                   onFileRemoved={(file) => { console.log(file); }}
                   onError={(error) => { console.log(error); }}
                   maxFileSize={104857600}
                   categories="14,15,16"
                   maxFiles={3}
                   onMaxFileSizeErrorCallback={function (files) {
                       alert("max size: 104857600");
                   }}
                   maxFilesErrorCallback={function () {
                       alert("Max Files: 1");
                   }}
               />
            </div>

        );
    }
}
```

### Demo

`npm run demo`

changelog.md
### Changelog
- 1.0.14
    - changed default fileAccept
    - added option fileAccept
- 1.0.12
    - added new options
        - onMaxFileSizeErrorCallback
        - maxFilesErrorCallback
    - updated react-resumablejs
- 1.0.11
    - updated react-resumablejs
- 1.0.10
    - added option maxFiles
- 1.0.9
    - added option categories. Now you can set Categories ID from Kaltura.
    - updated example.
- 1.0.8
    - updated react-resumablejs module.
- 1.0.7
    - updated Demo, now can login with Kaltura user/pass to get the KS.
    - updated react-resumablejs module.
    - added disableInput option to enable or disable the input file.
- 1.0.6
    - fixed demo.
    - added maxFileSize options ( in bytes ).
- 1.0.5
    - updated react resumablejs.
- 1.0.4
    - added module open as dev dependencies.
    - added option onFileRemoved.
    - changed onSuccess to onFileSuccess.
    - updated example.
- 1.0.1
    - updated react-resumablejs.

src/ReactKalturaResumableJs.js
import React from "react";
import axios from "axios";
import ReactResumableJs from "react-resumable-js";

const KALTURA_ADD = "/service/uploadToken/action/add";
const KALTURA_UPLOAD = "/service/uploadToken/action/upload";
const KALTURA_ADD_FROM_UPLOADED_FILE = "/service/media/action/addFromUploadedFile";

export default class ReactKalturaResumableJs extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            lastUploadToken: null
        };
    }

    _request = (path, queryString) => {

        let url = this.props.server + path + "?format=1&ks=" + this.props.ks + "&" + queryString;

        return new Promise(function (fulfill, reject) {
            axios.post(url)
                .then(fulfill)
                .catch(reject);
        });
    };

    _addMedia = (file, message) => {

        let path = KALTURA_ADD_FROM_UPLOADED_FILE;
        let queryString = "mediaEntry:name=" + encodeURIComponent(file.uniqueIdentifier) + "&mediaEntry:mediaType=1&uploadTokenId=" + this.state.lastUploadToken;

        if(this.props.categories.length > 0) {
            queryString+= '&mediaEntry:categoriesIds=' + this.props.categories;
        }

        this._request(path, queryString).then((response) => {
            if (typeof this.props.onFileSuccess === "function") {
                this.props.onFileSuccess(response.data);
            }
        }).catch((reject) => {
            if (typeof this.props.onError === "function") {
                this.props.onError(reject);
            }
        });
    };

    _uploadVideo = (file, resumable) => {

        file.bootstrap();

        let chunksize = this.props.chunksize;

        resumable.opts.target = this.props.server + KALTURA_UPLOAD;
        resumable.opts.fileParameterName = "fileData";
        resumable.opts.chunkSize = chunksize * 1024;
        resumable.opts.simultaneousUploads = this.props.simultaneousUploads;
        resumable.opts.testChunks = false;
        resumable.opts.throttleProgressCallbacks = 1;

        let queryString = "uploadToken:objectType=KalturaUploadToken" +
            "&uploadToken:fileName=" + encodeURIComponent(file.fileName) +
            "&uploadToken:fileSize=" + file.size;

        this._request(KALTURA_ADD, queryString)
            .then((response) => {

                this.setState({
                    lastUploadToken: response.data.id
                });

                resumable.opts.query = (file, chunk) => {
                    return {
                        format: 1,
                        ks: this.props.ks,
                        uploadTokenId: this.state.lastUploadToken,
                        resume: chunk.offset > 0 ? 1 : 0,
                        resumeAt: chunk.startByte,
                        finalChunk: chunk.offset + 1 == file.chunks.length ? 1 : 0,
                    };
                };

                resumable.upload();

            }).catch((reject) => {
            this.props.onError(reject);
        });
    };

    render() {

        return (
            <div>
                <ReactResumableJs
                    uploaderID={this.props.uploaderID}
                    dropTargetID="kalturaDropTarget"
                    filetypes={this.props.filetypes}
                    fileAccept={this.props.fileAccept}
                    fileAddedMessage={this.props.fileAddedMessage}
                    completedMessage={this.props.completedMessage}
                    service={null}
                    textLabel={this.props.textLabel}
                    disableDragAndDrop={true}
                    onFileSuccess={this._addMedia}
                    onFileAdded={this._uploadVideo}
                    onFileRemoved={this.props.onFileRemoved}
                    maxFileSize={this.props.maxFileSize}
                    disableInput={this.props.disableInput}
                    maxFiles={this.props.maxFiles}
                    onMaxFileSizeErrorCallback={this.props.onMaxFileSizeErrorCallback}
                    maxFilesErrorCallback={this.props.maxFilesErrorCallback}
                />
            </div>
        )
    }
}

ReactKalturaResumableJs.propTypes = {
    uploaderID: React.PropTypes.string.isRequired,
    filetypes: React.PropTypes.array,
    fileAccept: React.PropTypes.string,
    chunksize: React.PropTypes.number,
    simultaneousUploads: React.PropTypes.number,
    onError: React.PropTypes.func,
    onSuccess: React.PropTypes.func,
    onFileRemoved: React.PropTypes.func,
    textLabel: React.PropTypes.string,
    fileAddedMessage: React.PropTypes.string,
    completedMessage: React.PropTypes.string,
    maxFileSize: React.PropTypes.number,
    disableInput: React.PropTypes.bool,
    categories: React.PropTypes.string
};

ReactKalturaResumableJs.defaultProps = {
    filetypes: ['mp4'],
    fileAccept: "video/mp4,video/x-m4v,video/*,video/x-quicktime",
    chunksize: 1024,
    maxFileSize: 104857600,
    simultaneousUploads: 1,
    onError: (error) => { return error; },
    onFileSuccess: (success) => { return success;},
    onFileRemoved: (file) => { return file; },
    uploaderID: "video-uploader",
    textLabel: "Uploaded files",
    fileAddedMessage: "Started",
    completedMessage: "Complete!:",
    disableInput: false,
    categories: "",
    maxFiles: 20,
    onMaxFileSizeErrorCallback: null
};